wa/workloads/googlemaps/uiauto/app/src/main/java/com/arm/wa/uiauto/UiAutomation.java

```diff
@@ -0,0 +1,149 @@
+import android.os.Bundle;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import android.support.test.runner.AndroidJUnit4;
+
+import android.util.Log;
+import android.view.KeyEvent;
+
+// Import the uiautomator libraries
+import android.support.test.uiautomator.UiObject;
+import android.support.test.uiautomator.UiObjectNotFoundException;
+import android.support.test.uiautomator.UiScrollable;
+import android.support.test.uiautomator.UiSelector;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import com.arm.wa.uiauto.BaseUiAutomation;
+
+@RunWith(AndroidJUnit4.class)
+public class UiAutomation extends BaseUiAutomation {
+
+    protected Bundle parameters;
+    protected String packageID;
+    protected boolean offlineMode;
+
+    public static String TAG = "googlemaps";
+
+    @Before
+    public void initialize() throws Exception {
+        parameters = getParams();
+        packageID = getPackageID(parameters);
+        offlineMode = parameters.getBoolean("offline_mode");
+    }
+
+    @Test
+    public void setup() throws Exception {
+        setScreenOrientation(ScreenOrientation.NATURAL);
+        runApplicationSetup();
+    }
+
+    public void search(String query, String box) throws Exception {
+        UiObject searchBox = mDevice.findObject(new UiSelector().resourceId(packageID + box)
+                                                                .className("android.widget.EditText"));
+        if (!searchBox.waitForExists(uiAutoTimeout)) {
+            throw new UiObjectNotFoundException("Could not find search box.");
+        }
+        searchBox.click();
+
+        UiObject searchText = mDevice.findObject(new UiSelector().resourceId(packageID + "search_omnibox_edit_text")
+                                                                 .className("android.widget.EditText"));
+        searchText.click();
+        searchText.setText(query);
+    }
+
+    public void selectSearchResultContaining(String str) throws Exception {
+        UiObject match = mDevice.findObject(new UiSelector().textContains(str)
+                                                            .className("android.widget.TextView"));
+        if (!match.waitForExists(uiAutoTimeout)) {
+            throw new UiObjectNotFoundException("Could not find search result containing \"" + str + "\".");
+        }
+        match.click();
+    }
+
+    public void getDirectionsFromLocation() throws Exception {
+        UiObject directions = mDevice.findObject(new UiSelector().resourceId(packageID + "placepage_directions_button"));
+        directions.clickAndWaitForNewWindow(uiAutoTimeout);
+    }
+
+    @Test
+    public void runWorkload() throws Exception {
+        // Search for and select Cambridge train station
+        search("Cambridge Station CBG, England", "search_omnibox_text_box");
+        selectSearchResultContaining("[CBG]");
+
+        // Dismiss the tutorial dialog for place information if it appears
```

Comment:
Rather than commenting on what each code block does, please break them out into appropriately named functions and call them here. (e.g. see the implementation of googleplaybooks: https://github.com/ARM-software/workload-automation/blob/next/wa/workloads/googleplaybooks/uiauto/app/src/main/java/com/arm/wa/uiauto/googleplaybooks/UiAutomation.java)

This will make it easier to see the overall flow and to find the code for a particular action when it needs to be updated.

Reply:
Personally I'm not convinced that this will make the code easier to read, hence why I didn't write it this way, since in my mind this introduces unnecessary indirection (straight-line code is really easy to reason about!), but happy to make this change if you think it to be worthwhile.

Reply:
> straight-line code is really easy to reason about 

True, but breaking these blocks out into separate functions does not make the code any less straight-line -- there is still only a single path. And each of the code blocks is logically distinct from the other -- they do not need to be considered together to understand the code, so having them next to each other in the same function does not facilitate clarity.

On the other hand, in practice, maintaining the UI automation code comes down to making small changes to a specific operation in response to some UI change. Finding where in the code said operation takes place, especially for someone not already familiar with the automation flow, is made a lot easier if each operation is "folded" into its own function. 

> and I'm not re-using any of the code blocks here

Re-use is far from the only justification for functions.

Reply:
> Re-use is far from the only justification for functions.

Agreed, but what I was trying to get at is that I find the over-application of abstraction to be much more of a problem than the under-application in cases like this. That is, [I favour "style C" as described by John Carmack here](http://number-none.com/blow/john_carmack_on_inlined_code.html). In this case though we're talking about a tiny change in a tiny file so it doesn't really matter... I'll make some changes to clean it up a little and you can see what you think.

Reply:
He didn't really make a case for "style C" there, IMO. Most of his argument comes down to "long call chains are bad", and I agree with that; I mean, nobody wants [this](https://twitter.com/fioraaeterna/status/689908645476184068?lang=en). But that's not an argument for favouring style C over style B in a particular situation. Both of them have advantages and disadvantages, and in different situations, one might be more appropriate than the other.

In this case, what tends to happen, is:
- Workload fails
- You re-run and observe UI interactions
- You spot where it goes wrong
- You go to the automation source to fix the issue

Now, with style C, you have to scan through several screen-fulls of code to identify the part that needs fixing. With style B, you look at the outline of execution, and can immediately identify the relevant function (and yes, descriptive names are of paramount importance here), and can  jump to the relevant code right away.